File: TSP_Recursion.py
#r Using Recursion - O(n!) Time and O(n) Space

# Python program to find the shortest possible route
# that visits every city exactly once and returns to
# the starting point using memoization and bitmasking

#y tsp(curr, mask) = min(cost[curr][i] + tsp(i, mask ∣ (1<<i))), for all cities i that have not been visited yet.
#b curr is the current city in the tour.
#b mask represents the cities that have already been visited.
#b cost[curr][i] is the cost to travel from city curr to city i.
#b tsp(i, mask | (1 << i)) represents the cost of visiting the remaining cities in the new mask (after visiting city i), and continuing the tour from city i.

import sys

def totalCost(mask, pos, n, cost, path, memo):
  
    # Base case: if all cities are visited, return the
    # cost to return to the starting city (0)
    if mask == (1 << n) - 1:
        return cost[pos][0], [0]

    if (mask, pos) in memo:
        return memo[(mask, pos)]

    ans = sys.maxsize
    best_path = []

    # Try visiting every city that has not been visited yet
    for i in range(n):
        if (mask & (1 << i)) == 0: 
  
            # If city i is not visited, visit it and 
             #  update the mask
            sub_cost, sub_path = totalCost(mask | (1 << i), i, n, cost, path, memo)
            temp_cost = cost[pos][i] + sub_cost
            if temp_cost < ans:
                ans = temp_cost
                best_path = [i] + sub_path

    memo[(mask, pos)] = (ans, best_path)
    return memo[(mask, pos)]
 

def tsp(cost):
    n = len(cost)
    
    # Start from city 0, and only city 0 is visited 
    # initially (mask = 1)
    memo = {}
    total_cost, route = totalCost(1, 0, n, cost, [], memo)
    route = [0] + route  # prepend starting city
    return total_cost, route
 
if __name__ == "__main__":
    
    cost = [
        [0, 10, 100, 20, 56],
        [10, 0, 35, 25, 21],
        [100, 35, 0, 30, 12],
        [20, 25, 30, 0, 12],
        [56, 21, 12, 12, 0]
    ]

    city_names = ["Toko A", "Toko B", "Toko C", "Toko D", "Toko E"]

    result_cost, result_route = tsp(cost)
    print("Total cost:", result_cost)
    print("Route taken (index):", result_route)
    print("Route taken (named):", [city_names[i] for i in result_route])

    print("\nDetail perjalanan:")
    for i in range(len(result_route) - 1):
        a = result_route[i]
        b = result_route[i + 1]
        print(f"{city_names[a]} — {cost[a][b]} km → {city_names[b]}")

File: TSP_Recursion.java
// Java program to find the shortest possible route
// that visits every city exactly once and returns to
// the starting point using memoization and bitmasking

import java.util.*;

public class TSP_Recursion {

    static int[][] memo;
    static int[][] bestPath;
    static String[] cityNames;

    // Recursive TSP with memoization and path tracking
    static int totalCost(int mask, int pos, int n, int[][] cost) {
        // Base case: all cities visited, return to start
        if (mask == (1 << n) - 1) {
            return cost[pos][0];
        }

        if (memo[pos][mask] != -1) return memo[pos][mask];

        int ans = Integer.MAX_VALUE;
        int nextCity = -1;

        for (int i = 0; i < n; i++) {
            if ((mask & (1 << i)) == 0) {
                int newMask = mask | (1 << i);
                int temp = cost[pos][i] + totalCost(newMask, i, n, cost);
                if (temp < ans) {
                    ans = temp;
                    nextCity = i;
                }
            }
        }

        memo[pos][mask] = ans;
        bestPath[pos][mask] = nextCity;
        return ans;
    }

    // Reconstruct the route from bestPath table
    static List<Integer> reconstructPath(int n) {
        List<Integer> path = new ArrayList<>();
        int mask = 1;
        int pos = 0;
        path.add(pos);
        int steps = 0;

        while (true) {
            int next = bestPath[pos][mask];
            if (next == -1 || steps++ > n) break; // prevent infinite loop
            path.add(next);
            mask |= (1 << next);
            pos = next;
        }

        path.add(0); // return to starting city
        return path;
    }

    public static void main(String[] args) {
        int[][] cost = {
            { 0, 10, 100, 20, 56 },
            { 10, 0, 35, 25, 21 },
            { 100, 35, 0, 30, 12 },
            { 20, 25, 30, 0, 12 },
            { 56, 21, 12, 12, 0 }
        };

        cityNames = new String[] {
            "Toko A", "Toko B", "Toko C", "Toko D", "Toko E"
        };

        int n = cost.length;
        memo = new int[n][1 << n];
        bestPath = new int[n][1 << n];
        for (int[] row : memo) Arrays.fill(row, -1);
        for (int[] row : bestPath) Arrays.fill(row, -1);

        int resultCost = totalCost(1, 0, n, cost);
        List<Integer> route = reconstructPath(n);

        System.out.println("Total cost: " + resultCost);
        System.out.println("Route taken (index): " + route);
        System.out.print("Route taken (named): ");
        for (int i = 0; i < route.size(); i++) {
            System.out.print(cityNames[route.get(i)]);
            if (i != route.size() - 1) System.out.print(" → ");
        }

        System.out.println("\n\nDetail perjalanan:");
        for (int i = 0; i < route.size() - 1; i++) {
            int a = route.get(i);
            int b = route.get(i + 1);
            System.out.println(cityNames[a] + " - " + cost[a][b] + " km - " + cityNames[b]);
        }
    }
}
